### PR TITLE
commitlog: Add counters for segment waiting, create, delete and recycle

### PR DESCRIPTION
**Pull request by elcallio**
Because more counters == more fun.

Mainly to help QA diagnose run errors (in 4.5). Might apply relatively cleanly in a backport.

### CODE REVIEW

File: db/commitlog/commitlog.cc [hunk 1]
```diff
@@ -1378,6 +1383,15 @@ void db::commitlog::segment_manager::create_counters(const sstring& metrics_cate
...
                        sm::description("Holds the total number of bytes in internal memory buffers.")),
+
+        sm::make_gauge("waiting_for_segment", totals.waiting_for_segment,
+                       sm::description("Holds the number of segment allocations waiting for disk space to become available.")),
+
+        sm::make_derive("segment_files_created", totals.segment_files_created, sm::description("Counts a number of segment files created.")),
+
+        sm::make_derive("segment_files_recycled", totals.segment_files_recycled, sm::description("Counts a number of segment files recycled.")),
+        
+        sm::make_derive("segment_files_deleted", totals.segment_files_deleted, sm::description("Counts a number of segment files deleted.")),
```
Comment (bhalevy) [inline db/commitlog/commitlog.cc:1394]: nit: "a number" or "the number"?

**Review by bhalevy — APPROVED** [review]
LGTM (just a minor comment regarding the wording of the new counters' description)

File: db/commitlog/commitlog.cc (hunk 1 above)
Comment (elcallio) [reply db/commitlog/commitlog.cc]: Duh. Copy paste-error. Fixed.

**Comment by bhalevy** [comment]
@scylladb/scylla-maint please merge.
Next, we'd want to backport to 4.5 ASAP.

**Comment by avikivity** [comment]
Can you provide a private build for QA with this? I don't think these are generally useful.

**Comment by bhalevy** [comment]
If we're going in the private build path, then it better be based on branch-4.5 (@roydahan please confirm)

**Comment by elcallio** [comment]
I probably can, but in that case I think QA applying the patch to whatever tree they want to test is more likely to be the build wanted/delivered within reasonable time. Or is there a 4.5-capable-build-your-branch-into-qa-applicable-packages easily accessible in jenkins?

**Comment by roydahan** [comment]
Because the issue is easily reproducible also in master, I don't have to use 4.5.
However, I need rpms or  deb packages to test it.

**Comment by avikivity** [comment]
@elcallio cheat sheet for producing .deb:

1. apply your patches
2. change SCYLLA_BUILD to something and commit so it won't be mistaken for a clean build
3. `ninja release-dist`

**Comment by elcallio** [comment]
@avikivity that fails on fedora 34, can't create relocatable packages (`ldd build/dev/iotune: could not resolve libboost_program_options.so.1.69.0`) - so, one tries using dbuild: 
```
>./tools/toolchain/dbuild ./configure.py --disable-dpdk
Unable to find image 'scylladb/scylla-toolchain:fedora-34-20210731' locally
docker: Error response from daemon: missing signature key.
```
And at this point I am out of ideas, and not much closer to a usable rpm/deb set.

**Comment by avikivity** [comment]
> @avikivity that fails on fedora 34, can't create relocatable packages (`ldd build/dev/iotune: could not resolve libboost_program_options.so.1.69.0`)

You likely have a very old iotune (Fedora 34 has boost 1.75). Unfortunately cmake dependency tracking is bad, so it didn't try to rebuild it. Wipe build/release/sestar and build again.

> - so, one tries using dbuild:
> 
> ```
> >./tools/toolchain/dbuild ./configure.py --disable-dpdk
> Unable to find image 'scylladb/scylla-toolchain:fedora-34-20210731' locally
> docker: Error response from daemon: missing signature key.
> ```
> 

This looks like a local docker storage misconfiguration, or perhaps an outdated docker install. Upgrade your docker and/or wipe its local storage. Better, drop docker and install podman, it does the same thing without a daemon.

> And at this point I am out of ideas, and not much closer to a usable rpm/deb set.

**Comment by bhalevy** [comment]
Is this PR still applicable or can it be closed?

**Comment by mykaul** [comment]
> Is this PR still applicable or can it be closed?

@elcallio - specifically, with the future changes of the commitlog (compression, some rewrite, etc.) looks relevant to me.

**Comment by elcallio** [comment]
I'd very much like to fold this into the refactor branch at least.

**Pull request closed without merging**